File: 0021-merge-two-sorted-lists/0021-merge-two-sorted-lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    
    public ListNode solve(ListNode list1, ListNode list2){
        ListNode cur = list1;
        ListNode next = list1.next;
        if(next == null){
            cur.next = list2;
        }
        ListNode temp = list2;
        ListNode next2 = null;
        while(temp != null && next != null){
            
            if(cur.val <= temp.val && temp.val <= next.val){
                cur.next = temp;
                next2 = temp.next;
                temp.next = next;
                cur = temp;
                temp = next2;
            }
            else{
                cur = next;
                next = next.next;
                if(next == null){
                    cur.next = temp;
                    return list1;
                }
            }
            
        }
        return list1;
    }
    
    public ListNode mergeTwoLists(ListNode list1, ListNode list2) {
        
        if(list1 == null)
            return list2;
        if(list2 == null)
            return list1;
        
        if(list1.val <= list2.val){
            return solve(list1, list2);
        }
        
        else{
            return solve(list2, list1);
        } 
    }
}

File: 0021-merge-two-sorted-lists/0021-merge-two-sorted-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
        if(!list1)
            return list2;
        if(!list2)
            return list1;

        ListNode* prev1 = NULL;
        ListNode* temp1 = list1;
        ListNode* temp2 = list2;
        while(temp1 && temp2){
            if(temp1->val <= temp2->val){
                prev1 = temp1;
                temp1 = temp1->next;
            }
            else{
                if(prev1)
                    prev1->next = temp2;
                else
                    list1 = temp2;
                ListNode* next2 = temp2->next;
                temp2->next = temp1;
                prev1 = temp2;
                temp2 = next2;
            }
        }
        while(temp2){
            prev1->next = temp2;
            prev1 = temp2;
            temp2 = temp2->next;
        }
        return list1;
    }
};